CI job running tests with queuing on

gjoseph92 opened the pull request:
Adds a CI job, only for Ubuntu and py3.9, running the test suite with `worker-saturation: 1.0` (instead of `inf`).

Also adds a `@pytest.mark.oversaturate_only` marker, which automatically skips marked tests if queuing is enabled.

A surprisingly small number of tests actually failed when queuing was turned on. I've updated the ones that did as needed. I've tried as much as possible to just make the tests agnostic to queuing. In a few cases though, I've needed to add explicit conditional logic. And in a couple cases, I've marked tests to be skipped when queuing is enabled, since they may not make sense.

Unfortunately there are three tests that are flaky under queuing, which could indicate an actual bug. They're skipped for now, but should be investigated.

Closes https://github.com/dask/distributed/issues/6631

- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

crusaderky commented on .github/workflows/tests.yaml:
```diff
@@ -107,6 +116,11 @@ jobs:
+        if: ${{ matrix.queuing == 'queue' }}
+        shell: bash -l {0}
+        run: export DASK_DISTRIBUTED__SCHEDULER__WORKER_SATURATION=1.0
+
```
I thought that environment variables are reset at every step of the github action?

crusaderky reviewed (changes requested):
Instead of this new `oversaturate_only` mark, couldn't you simply have

```python
OVERSATURATION = math.isinf(dask.config.get("distributed.scheduler.worker-saturation"))
@pytest.mark.flaky(not OVERSATURATION, reason="flaky on MacOS")
```
Tests that don't make sense with queueing should instead force 
```python
config={"distributed.scheduler.worker-saturation": math.inf}
```
(some already do)

fjetter commented on .github/workflows/tests.yaml:
```diff
@@ -44,8 +53,8 @@ jobs:
     env:
-      TEST_ID: ${{ matrix.os }}-${{ matrix.python-version }}-${{ matrix.partition-label }}
-      # TEST_ID: ${{ matrix.os }}-${{ matrix.python-version }}-${{ matrix.partition-label }}-${{ matrix.run }}
+      TEST_ID: ${{ matrix.os }}-${{ matrix.python-version }}-${{ matrix.partition-label }}-${{ matrix.queuing }}
```
I think this renaming will break some logic in our test report generation.

https://github.com/dask/distributed/blob/acf607832c7191cc496a9b4a81760170de85062c/continuous_integration/scripts/test_report.py#L146-L152
https://github.com/dask/distributed/blob/acf607832c7191cc496a9b4a81760170de85062c/continuous_integration/scripts/test_report.py#L353-L357

I introduced this in https://github.com/dask/distributed/pull/6837 

Worst case, we remove the more specific job URL again (we knew from the start this is very brittle)

fjetter replied:
see https://github.com/dask/distributed/pull/6837/files#r939182529

fjetter replied:
I don't think this should block this PR

gjoseph92 commented:
* Removed the `oversaturate_only` mark; just using `pytet.mark.skipif` and `config={"distributed.scheduler.worker-saturation": math.inf}` as necessary
* Fixed setting the environment variable in the test step
* Updated test report script to handle and display results from the `queue` job:

    ![visualization](https://user-images.githubusercontent.com/3309802/188240129-aa900b7d-c930-4e62-9ea2-20447cc38bb0.png)

Ready for re-review and merge.

gjoseph92 commented:
- https://github.com/dask/distributed/issues/6999

crusaderky commented:
Thank you!

jrbourbeau commented:
Just to confirm, is the new CI job added here for testing purposes and is planned to be removed in the future, or is this intended as a long-term change?

crusaderky commented:
Long term, we'll want to enable scheduler-side queueing by default.
When that happens, we'll need to revisit the unit tests to have selected few tests with queueing off (like it already happens with all other config toggles) and revert this PR.
Before that, all tests that are flaky only when queueing is on need to be addressed.
